src/ghga_datasteward_kit/s3_upload/uploader.py

```diff
@@ -53,28 +127,41 @@ async def schedule(self, fn: Coroutine[Any, Any, None]):
 
     async def gather(self):
         """Await all running tasks"""
-        await asyncio.gather(*self._tasks)
+        try:
+            await asyncio.gather(*self._tasks)
+        except PartUploadError:
+            # root cause should always be PartUploadError if `send_part` is scheduled as task
+            # cancel still running tasks and propagate exception to caller
+            for task in self._tasks:
+                if not task.done():
+                    task.cancel()
+            raise
 
 
 class ChunkedUploader:
     """Handler class dealing with upload functionality"""
 
-    def __init__(
+    def __init__(  # noqa: PLR0913
         self,
+        *,
         input_path: Path,
         alias: str,
         config: LegacyConfig,
         unencrypted_file_size: int,
-        storage_cleaner: StorageCleaner,
+        encryptor: Encryptor,
+        decryptor: Decryptor,
     ) -> None:
         self.alias = alias
         self.config = config
         self.input_path = input_path
-        self.encryptor = Encryptor(self.config.part_size)
+        self.encryptor = encryptor
+        self.decryptor = decryptor
         self.file_id = str(uuid4())
         self.unencrypted_file_size = unencrypted_file_size
         self.encrypted_file_size = 0
-        self._storage_cleaner = storage_cleaner
+        self.retry_handler = configure_retries(config)
+        self._in_sequence_part_number = 1
+        self._semaphore = asyncio.Semaphore(config.client_max_parallel_transfers)
```

Comment:
Need to check, but at least semaphore needs to be shared and thus instantiated outside of `send_part`, but could still be passed into it instead.

Reply:
I'd leave the semaphore where it is right now, I find it easier to reason about it as shared state and not something the caller of `send_part` should worry about. As already mentioned, there's no way to just declare it inside the function and it would need to pass a method boundary instead.

Wrt `retry_handler`: There's some repetitive configuration involved if that was moved to its actual call site and reconfiguration for each file part sent seems a bit excessive.

Reply:
Makes sense, thanks for the explanation.